res/controllers/Arturia-KeyLab-Mk1-scripts.js

```diff
@@ -0,0 +1,1007 @@
+                key: "pregain",
+                softTakeover: softTakeover,
+            });
+            this.highFilterGainPot = new components.Pot({
+                group: "[EqualizerRack1_[Channel1]_Effect1]",
+                midi: [0xB0 + midiChannel, 0x47],
+                key: "parameter3",
+                softTakeover: softTakeover,
+            });
+            this.midFilterGainPot = new components.Pot({
+                group: "[EqualizerRack1_[Channel1]_Effect1]",
+                midi: [0xB0 + midiChannel, 0x4C],
+                key: "parameter2",
+                softTakeover: softTakeover,
+            });
+            this.lowFilterGainPot = new components.Pot({
+                group: "[EqualizerRack1_[Channel1]_Effect1]",
+                midi: [0xB0 + midiChannel, 0x4D],
+                key: "parameter1",
+                softTakeover: softTakeover,
+            });
+            this.quickEffectPot = new components.Pot({
+                group: "[QuickEffectRack1_[Channel1]]",
+                midi: [0xB0 + midiChannel, 0x5D],
+                key: "super1",
+                softTakeover: softTakeover,
+            });
+
+            // Other encoders
+            this.valueEncoder = new components.Encoder({
+                group: "[Channel1]",
+                midi: [0xB0 + midiChannel, 0x72], // Value encoder
+                input: function(_channel, _control, value, _status, _group) {
+                    if (!this.inKey) {
+                        // If we haven't selected a parameter yet, don't do
+                        // anything.
+                        return;
+                    }
+                    const param = this.inGetParameter();
+                    if (value === 0x3F) {
+                        if (this.relative) {
+                            this.inSetParameter(-1);
+                            return;
+                        }
+                        const newParam = param - 1;
+                        if (
+                            this.min !== undefined && this.max !== undefined
+                            && newParam < this.min
+                        ) {
+                            this.inSetParameter(this.max);
+                        } else {
+                            this.inSetParameter(param - 1);
+                        }
+                    } else if (value === 0x41) {
+                        if (this.relative) {
+                            this.inSetParameter(1);
+                            return;
+                        }
+                        const newParam = param + 1;
+                        if (
+                            this.max !== undefined && this.min !== undefined
+                            && newParam > this.max
+                        ) {
+                            this.inSetParameter(this.min);
+                        } else {
+                            this.inSetParameter(param + 1);
+                        }
+                    }
```

Comment:
this looks very convoluted. I also think there is a bug in the clamping logic (if `newParam > this.max` but then you `this.inSetParameter(this.min);` and vice versa). 
```suggestion
                    value = value - 0x40;
                    if (this.relative) {
                        this.inSetParameter(value);
                        return;
                    }
                    const newParam = param + value;
                    if (this.min !== undefined && this.max !== undefined) {
                        this.inSetParameter(script.clamp(newParam, this.min, this.max));
                    } else {
                        this.inSetParamter(newParam);
                    }
```
for this to work, add a `script.clamp = function(val, low, high) {return Math.min(high, Math.max(low, val));}` to `common-controller-scripts.js`.

Reply:
I had forgotten about this; I didn't realize it was just 2's compliment values that are being used to indicate velocity when I did this. I've since written a few more and realized why it's 0x41 or what not. I'll change this over to the regular implementation.

It might be worth having a simple way to decode and scale the value in the `script` namespace too (that way those writing XML mappings have access to it for infinite encoders, which right now just aren't supported in XML-only stuff).

Reply:
I did make some changes to make this work regardless of how fast you turn the encoder, but the logic ended up being basically the same. I had forgotten that this isn't clamping, it's wrapping around on the ends between min and max (ie. if we sort by library columns and we hit last_played (31) we want to sort by artist (1) next, not an undefined value). I suppose we could add some sort of wrap function too, but it seems less useful so I left the logic as-is. I'm still open to simplification, of course, but this seems the most readable way for now.